### PR TITLE
[fuchsia] Output Fuchsia OS crash data to dashboard.

### PR DESCRIPTION
**Pull request by flowerhack**
Initially I promised this would include both crash data & a symbolized
stacktrace; however, some hurdles wrt symbolized output on the Fuchsia
end of things has made me think it may be cleaner/faster to go ahead and
get this "hello world" out there.

When Clusterfuzz finds a crash, a testcase will appear on the
Clusterfuzz dashboard, e.g. like so:

```	[Environment] ASAN_OPTIONS="redzone=32:print_summary=1:handle_sigill=1:allocator_release_to_os_interval_ms=500:print_suppressions=0:strict_memcmp=1:allow_user_segv_handler=0:allocator_may_return_null=1:use_sigaltstack=1:handle_sigfpe=1:handle_sigbus=1:detect_stack_use_after_return=1:alloc_dealloc_mismatch=0:detect_leaks=1:print_scariness=1:max_uar_stack_size_log=16:handle_abort=1:check_malloc_usable_size=0:quarantine_size_mb=64:detect_odr_violation=0:symbolize=0:handle_segv=1:fast_unwind_on_fatal=0"
[Command line] python /usr/local/google/home/flowerhack/june13-2/clusterfuzz/src/python/bot/fuzzers/libFuzzer/launcher.py libFuzzer dummy_fuzzer_name
+----------------------------------------Release Build Stacktrace----------------------------------------+
Warning: Permanently added '[localhost.corp.google.com]:41105' (ED25519) to the list of known hosts.
INFO: Seed: 3708787003
INFO: Loaded 1 modules   (9 inline 8-bit counters): 9 [0x482cdb589060, 0x482cdb589069),
INFO: Loaded 1 PC tables (9 PCs): 9 [0x482cdb589070,0x482cdb589100),
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
==33853== ERROR: libFuzzer: deadly signal
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 1 ChangeByte-; base unit: 1e66b44810d66b9f096d6a5b1344d65f45401873
0x48,0x49,0x21,
HI!
artifact_prefix='data'; Test unit written to datacrash-7a8dc3985d2a90fb6e62e94910fc11d31949c348
Base64: SEkh```

Note that this won't include nice things (yet) like: a downloadable
crash, symbolized output, etc.

However, following the instructions at
https://github.com/google/fuzzer-test-suite/blob/master/tutorial/libFuzzerTutorial.md
*should* be sufficient for root-causing bugs, even if it's not as
convenient, so I think this is a reasonable first step for us to take.

I'll attach some screenshots of the current output, as well as a list of
"future tasks" we'll plan for in the future to further prettify this
workflow.

Feedback welcome, and thanks as always for reviewing!

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
>`artifact_prefix='data'` 

is this supposed to be `artifact_prefix='data/'`?

I think this approach will behave unexpectedly if a fuzzer has more than one crash at a time.

**Comment by flowerhack** [comment]
> > `artifact_prefix='data'`
> 
> is this supposed to be `artifact_prefix='data/'`?
> 
> I think this approach will behave unexpectedly if a fuzzer has more than one crash at a time.

@nopsledder ?

**Comment by oliverchang** [comment]
Try applying this patch: https://gist.github.com/oliverchang/c69ad130ae57ccd47ad884a5f14b7b0e 

It should get rid of the necessity of most of the hacks here that deal with testcase_file_path not existing.

**Comment by oliverchang** [comment]
I've updated the patch above, and now the crash appears on the dashboard again. There is still a lot of cleanup that can be done though.

File: src/python/bot/fuzzers/libFuzzer/launcher.py [hunk 1]
```diff
@@ -860,9 +861,13 @@ def main(argv):
   new_testcases_directory = create_corpus_directory('new')
 
   # Get list of corpus directories.
-  corpus_directories = get_corpus_directories(
-      corpus_directory, new_testcases_directory, fuzzer_path,
-      fuzzing_strategies, minijail_chroot)
+  # TODO(flowerhack): Implement this to handle corpus sync'ing.
+  if environment.platform() != 'FUCHSIA':
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libFuzzer/launcher.py]: nit:  invert the condition here so we compare == 'FUCHSIA'

File: src/python/bot/tasks/fuzz_task.py [hunk 2]
```diff
@@ -1285,6 +1284,10 @@ def execute_task(fuzzer_name, job_type):
   # If yes, bail out.
   logs.log('Checking for bad build.')
   crash_revision = environment.get_value('APP_REVISION')
+  # TODO(flowerhack): Remove this exceptional case once Fuchsia understands
```
Comment (oliverchang) [inline src/python/bot/tasks/fuzz_task.py]: nit: can we set APP_REVISION instead in FuchsiaBuild of adding this hack?

File: src/python/crash_analysis/crash_result.py [hunk 3]
```diff
@@ -39,6 +40,12 @@ def get_crash_time(self):
 
   def get_symbolized_data(self):
     """Compute symbolized crash data if necessary or return cached result."""
+    # TODO(flowerhack): Change here to support richer crash reports for Fuchsia.
+    # (Right now we just grab the output.)
+    if environment.platform() == 'FUCHSIA':
```
Comment (oliverchang) [inline src/python/crash_analysis/crash_result.py]: does removing this hack break anything? if so, what breaks?

File: src/python/crash_analysis/crash_result.py [hunk 4]
```diff
@@ -48,6 +55,10 @@ def get_symbolized_data(self):
 
   def get_unsymbolized_data(self):
     """Compute unsymbolized crash data if necessary or return cached result."""
+    # TODO(flowerhack): Change here to support richer crash reports for Fuchsia.
```
Comment (oliverchang) [inline src/python/crash_analysis/crash_result.py]: ditto for rest of these. it seems a little too messy to do these.

File: src/python/fuzzing/testcase_manager.py [hunk 5]
```diff
@@ -377,6 +377,10 @@ def convert_dependency_url_to_local_path(url):
 
 def _get_testcase_time(testcase_path):
   """Returns the timestamp of a testcase."""
+  # TODO(flowerhack): When we teach CF to find Fuchsia testcase paths,
+  # change this.
+  if environment.platform() == 'FUCHSIA':
```
Comment (oliverchang) [inline src/python/fuzzing/testcase_manager.py]: not sure if this is still necessary anymore, not that we have proper paths.

File: src/python/fuzzing/testcase_manager.py [hunk 6]
```diff
@@ -437,6 +441,10 @@ def run_testcase_and_return_result_in_queue(crash_queue,
     # Run testcase and check whether a crash occurred or not.
     return_code, crash_time, output = run_testcase(thread_index, file_path,
                                                    gestures, env_copy)
+    # TODO(flowerhack): Update this once Fuchsia knows how to properly extract
+    # crash time?
+    if crash_time is None:
```
Comment (oliverchang) [inline src/python/fuzzing/testcase_manager.py]: is this necessary? CF should automatically get the crash time based on observing when the process crashes.

File: src/python/metrics/fuzzer_logs.py [hunk 7]
```diff
@@ -89,6 +89,9 @@ def upload_to_logs(bucket_name,
   Returns:
     The path of the uploaded file and whether the uploaded succeeded.
   """
+  #TODO(flowerhack): Implement this for log support.
+  if environment.platform() == 'FUCHSIA':
```
Comment (oliverchang) [inline src/python/metrics/fuzzer_logs.py]: is this necessary? what breaks?

File: src/python/fuzzing/testcase_manager.py [hunk 8]
```diff
@@ -855,7 +863,9 @@ def get_command_line_for_application(file_to_run='',
                                           testcase_file_url)
 
   # TODO(flowerhack): If we'd like blackbox fuzzing support for Fuchsia, here's
-  # where to add in our app's launch command.
+  # where to add in our app's launch command, instead of this placeholder.
+  if plt == 'FUCHSIA':
+    command += ' dummy_fuzzer_name'
```
Comment (oliverchang) [inline src/python/fuzzing/testcase_manager.py]: what is this needed for?

**Comment by flowerhack** [comment]
@oliverchang ; if you have any ideas on why the test is still broken, i'm all ears!

**Comment by oliverchang** [comment]
Looks like the issue is that multiprocessing's map_async doesn't handle exceptions very well -- they don't include the traceback, making this a bit difficult to debug... still digging into this one.

**Comment by oliverchang** [comment]
If you rebase on top of https://github.com/google/clusterfuzz/pull/587, you should be able to see the actual errors:

```
ERROR: test_fuzz_crash (tests.core.bot.fuzzers.libFuzzer.libfuzzer_launcher_integration_test.TestLauncherMinijail)
Step #3: ======================================================================
Step #3: 
Step #3: ValueError: Unable to configure handler 'handler': [Errno 2] No such file or directory: '/workspace/bot/logs/run_fuzzer.log'
Step #3: '%r: %s' % (name, e))
```

It seems that something is incorrectly deleting the bot logs dir.

File: bot/logs/README.md [hunk 9]
```diff
@@ -1,27 +0,0 @@
-Placeholder for log files generated by the bots.
```
Comment (oliverchang) [inline bot/logs/README.md]: This is the reason for the test failure. Please restore this file.

**Comment by flowerhack** [comment]
Hey Oliver!

Sorry for the delay replying to this--I misread my Github notifications and thought this had gotten merged.

Fixed the two things you mentioned last round; lmk if things are good now!

**Pull request merged**